### PR TITLE
Add `close_to` param for specifying upfront shutdown script

### PR DESCRIPTION
**Pull request by niftynei**
Allow users to specify the `upfront_shutdown_script` via a new parameter on `fundchannel_start`.

### CODE REVIEW

File: lightningd/opening_control.c [hunk 1]
```diff
@@ -1078,6 +1092,7 @@ static struct command_result *json_fund_channel_start(struct command *cmd,
 			   p_req("amount", param_sat, &amount),
 			   p_opt("feerate", param_feerate, &feerate_per_kw),
 			   p_opt_def("announce", param_bool, &announce_channel, true),
+			   p_opt("close_to", param_bitcoin_address, &fc->our_upfront_shutdown_script),
```
[the remark this answers is not in the document]
Comment (darosior) [reply lightningd/opening_control.c]: I think it would made VALGRIND tests pass

**Comment by justinmoon** [comment]
Concept ACK. This would be really useful.

**Comment by trueptolemy** [comment]
https://github.com/ElementsProject/lightning/blob/a91b454749f43e5eea414ad0a3cb5f66adc9b666/lightningd/peer_control.c#L1459-L1475
Here's a `FIXME` that should be considered. If we set `close_to` in `fundchannel_start`, we must fail the `close` command if the `destination` doesn't equal to `close_to` address.

And here's should be another check:
https://github.com/ElementsProject/lightning/blob/bd55f6d9408d7928852637e9581a3d9d5107111f/channeld/channeld.c#L2791-L2794

**Comment by rsbondi** [comment]
> If we set close_to in fundchannel_start, we must fail the close command if the destination doesn't equal to close_to address.

I ran into an issue with this when testing, got an unknown random P2WSH address receiving the funds rather than `close_to` or `destination`, so needs a fix, but maybe `destination` should override the `close_to`?   But failing would be acceptable.

**Comment by trueptolemy** [comment]
> random P2WSH address

Maybe it is the internal wallet address? Have you executed only one `close` command? 

> but maybe destination should override the close_to?

I think if we set `close_to`(and it was accepted by the remote), it has the highest priority(because **in spec, `close` must accept upfront shutdown script**). Or if we set the `destination` of `close`, it has the second priority. Finally, the internal wallet address has the lowest priority.

**Comment by rsbondi** [comment]
> Maybe it is the internal wallet address

those are P2WPKH or P2SH

> Have you executed only one close command and specified the destination?

`close` with `destination` works as expected when I open without `close_to`

spec rules, so fail seems best.

**Comment by trueptolemy** [comment]
> those are P2WPKH or P2SH

Yeah, they're p2wpkh

File: lightningd/opening_control.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (niftynei) [reply lightningd/opening_control.c]: `p_opt_def` sets the variable to `NULL` by default; the problem was in the deprecated codepath. :)

**Comment by niftynei** [comment]
> Needs to go into db, I think, and we make sure it's the same before trying shutdown.

There's already a database field for it in `channels`: local_shutdown_script. It plugs directly into that.

> I ran into an issue with this when testing, got an unknown random P2WSH address receiving the funds rather than close_to or destination, so needs a fix, but maybe destination should override the close_to? But failing would be acceptable.

The `P2SH` script that you got should be the `unilateral` closing address. In the version of this that you tested, attempting to issue a 'close' command to a destination different than what was specified upfront results in the peer throwing an error and the channel closing to a unilateral output script.

I changed this so that if you attempt to called `close [destination]` with a destination that is not the same as  the upfront shutdown script you get an RPC error. One side effect of this change is that you'll no longer be able to call `close` mutliple times with different destinations, but this is a reasonable tradeoff.

> Here's a FIXME that should be considered. If we set close_to in fundchannel_start, we must fail the close command if the destination doesn't equal to close_to address.
> And here's should be another check:

regarding this second check, we check before we call into this; checking here would be redundant.

thanks for the reviews :+1:

File: doc/lightning-fundchannel_start.7 [hunk 2]
```diff
@@ -1,9 +1,9 @@
 .TH "LIGHTNING-FUNDCHANNEL_START" "7" "" "" "lightning-fundchannel_start"
 .SH NAME
-lightning-fundchannel_start - Command for initiating channel establishment for a lightning channel
+
```
Comment (rustyrussell) [inline doc/lightning-fundchannel_start.7]: How did this get corrupted?  Do you need to update your mrkd?

File: doc/lightning-fundchannel_start.7 [hunk 3]
```diff
@@ -26,6 +26,11 @@ commitment transactions\.
 \fIannounce\fR whether or not to announce this channel\.
 
 
+\fIclose_to\fR is a Bitcoin address to which the channel funds should be sent to
+on close\. Only valid if both peers have negotiated \fBoption_upfront_shutdown_script\fR\.
+Returns \fBusing_close_to\fR set to true if it will\.
```
Comment (rustyrussell) [inline doc/lightning-fundchannel_start.7]: Prefer just returning the actual shutdown script as "close_to" iff it's used?

File: doc/lightning-fundchannel_start.7 (hunk 2 above)
Comment (niftynei) [reply doc/lightning-fundchannel_start.7]: I did in fact need to update `mrkd`

**Comment by rustyrussell** [comment]
I've collapsed these fixes back into two commits, and am testing locally now:

```
E            - Node /tmp/ltests-msvtc38j/test_closing_specified_destination_1/lightning-1/ has memory leaks: [
E               {
E                   "backtrace": [
E                       "ccan/ccan/tal/tal.c:437 (tal_alloc_)",
E                       "ccan/ccan/tal/tal.c:466 (tal_alloc_arr_)",
E                       "bitcoin/script.c:294 (scriptpubkey_p2wpkh)",
E                       "lightningd/peer_control.c:190 (p2wpkh_for_keyidx)",
E                       "lightningd/peer_control.c:1458 (json_close)",
E                       "lightningd/jsonrpc.c:645 (parse_request)",
E                       "lightningd/jsonrpc.c:743 (read_json)",
E                       "ccan/ccan/io/io.c:59 (next_plan)",
E                       "ccan/ccan/io/io.c:407 (do_plan)",
E                       "ccan/ccan/io/io.c:417 (io_ready)",
E                       "ccan/ccan/io/poll.c:445 (io_loop)",
E                       "lightningd/io_loop_with_timers.c:24 (io_loop_with_timers)",
E                       "lightningd/lightningd.c:859 (main)"
E                   ],
E                   "label": "bitcoin/script.c:294:u8[]",
E                   "parents": [
E                       "lightningd/channel.c:190:struct channel"
E                   ],
E                   "value": "0x5645bf695b88"
```

**Comment by rustyrussell** [comment]
Rebased fixups back into two patches (no changes), then added a trivial fix on top, found by leak detection.

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/pull/3153/commits/9542e13d4988cabb83b3553a43a7b9cd505d61e5

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3153/commits/3a90f33e9faf7a408cd8595da7114d0ab7695c43

**Pull request merged**